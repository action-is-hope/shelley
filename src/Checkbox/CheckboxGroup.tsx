"use client";
import { Ref, forwardRef, ReactElement } from "react";
import type { AriaCheckboxGroupProps } from "@react-types/checkbox";
import type { CheckboxProps } from ".";
import { CheckboxGroupContext } from ".";
import type { Orientation } from "src/typings/shared-types";
import { useCheckboxGroupState } from "react-stately";
import { useCheckboxGroup } from "react-aria";
import { Field, FieldProps } from "../Field";
import { st, classes } from "./checkboxGroup.st.css";

export interface CheckboxGroupProps
  extends Omit<AriaCheckboxGroupProps, "excludeFromTabOrder" | "errorMessage">,
    Omit<
      FieldProps,
      | "label"
      | "startAdornment"
      | "endAdornment"
      | "disableLabelTransition"
      | "hasValue"
      | "variant"
    > {
  /** Custom className */
  className?: string;
  /**
   * Position of the label.
   * @default "over"
   */
  labelPosition?: "top" | "side" | "hidden";
  /**
   * The Checkboxes contained within the CheckboxGroup.
   */
  children: ReactElement<CheckboxProps> | ReactElement<CheckboxProps>[];
  /**
   * The axis the checkboxes should align with.
   * @default 'vertical'
   */
  orientation?: Orientation;
}

function CheckboxGroup(props: CheckboxGroupProps, ref?: Ref<HTMLDivElement>) {
  const {
    className: classNameProp,
    description,
    isDisabled,
    errorMessage,
    isInvalid,
    label,
    labelPosition = "top",
    orientation = "vertical",
    vol,
    children,
    "data-id": dataId,
  } = props;

  const state = useCheckboxGroupState(props);
  const { labelProps, groupProps, descriptionProps, errorMessageProps } =
    useCheckboxGroup(props, state);

  return (
    <Field
      {...{
        isDisabled,
        errorMessage,
        isInvalid,
        errorMessageProps,
        inputContainerProps: { ...groupProps, className: classes.group },
        description,
        descriptionProps,
        label,
        labelPosition,
        labelProps,
        vol,
        variant: false,
        ref,
        "data-id": dataId,
      }}
      className={st(classes.root, { orientation }, classNameProp)}
    >
      <>
        <CheckboxGroupContext.Provider value={state}>
          {children}
        </CheckboxGroupContext.Provider>
      </>
    </Field>
  );
}
CheckboxGroup.displayName = "CheckboxGroup";

const _CheckboxGroup = forwardRef(CheckboxGroup);
export { _CheckboxGroup as CheckboxGroup };
