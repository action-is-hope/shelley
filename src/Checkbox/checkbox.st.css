@st-namespace "Checkbox";
@st-import InputLabel from "../Label/label.st.css";

.root {
  -st-states:
    isDisabled,
    isFocusVisible,
    isIndeterminate,
    validationState(enum(valid, invalid)),
    size(number);
  -st-extends: InputLabel;
}

.inputContainer {}

.input {}
