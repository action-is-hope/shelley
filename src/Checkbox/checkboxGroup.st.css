@st-namespace "CheckboxGroup";

@st-import Field from "../Field/field.st.css";

.root {
  -st-extends: Field;
  -st-states: orientation(enum(vertical, horizontal));
}

.group {
  display: flex;
}

.root:orientation(vertical) .group {
  flex-direction: column;
}

.root:orientation(horizontal) .group {
  flex-flow: wrap;
}
