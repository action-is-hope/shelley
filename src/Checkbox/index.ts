export * from "./Checkbox";
export * from "./CheckboxGroup";
export * from "./context";
