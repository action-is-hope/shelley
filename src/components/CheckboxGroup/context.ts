import type { CheckboxGroupState } from "@react-stately/checkbox";
import React from "react";

export const CheckboxGroupContext =
  React.createContext<CheckboxGroupState | null>(null);
