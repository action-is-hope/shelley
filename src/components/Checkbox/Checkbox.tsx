import React, { forwardRef, useRef, useContext } from "react";
import { useToggleState } from "react-stately";
import { useCheckbox, useCheckboxGroupItem } from "react-aria";
import type { AriaCheckboxProps } from "@react-types/checkbox";
import { mergeRefs, mergeProps } from "@react-aria/utils";
import { useFocusRing } from "react-aria";
import type { Volume, AlignPos, ComponentBase } from "../types";
import Label from "../Label/Label";
import { CheckboxGroupContext } from "../CheckboxGroup/context";
/* = Style API. */
import { st, classes } from "./checkbox.st.css";

export interface CheckboxProps extends AriaCheckboxProps, ComponentBase {
  className?: string;
  /** The position of the label relative to the label. */
  inputPosition?: AlignPos;
  /** Visually hide the label so it is still accessible to assistive technologies. */
  visuallyHideLabel?: boolean;
  /** How 'loud' should this input be? */
  vol?: Volume;
}

const Checkbox = forwardRef(
  (props: CheckboxProps, ref?: React.Ref<HTMLInputElement>) => {
    const {
      className: classNameProp,
      children,
      validationState,
      visuallyHideLabel,
      inputPosition,
      isIndeterminate,
      vol = 1,
      isDisabled,
      "data-id": dataId,
    } = props;
    const localRef = useRef(null);
    const groupState = useContext(CheckboxGroupContext);

    const { inputProps } = groupState
      ? // eslint-disable-next-line react-hooks/rules-of-hooks
        useCheckboxGroupItem(
          {
            ...props,
            // Value is optional for standalone checkboxes, but required for CheckboxGroup items;
            // it's cast explicitly here to avoid typescript error.
            value: props.value as string,
            // Only pass isRequired and validationState to react-aria if they came from
            // the props for this individual checkbox, and not from the group via context.
            // isRequired: originalProps.isRequired,
            // validationState: originalProps.validationState
          },
          groupState,
          localRef
        )
      : // eslint-disable-next-line react-hooks/rules-of-hooks
        useCheckbox(props, useToggleState(props), localRef);
    const { isFocusVisible, focusProps } = useFocusRing();

    const classNames = st(
      classes.root,
      {
        isDisabled,
        isFocusVisible,
        isIndeterminate,
        validationState,
        vol: vol ? vol : undefined,
      },
      classNameProp
    );

    const inputControl = (
      <span className={classes.inputContainer}>
        <input
          className={classes.input}
          {...mergeProps(inputProps, focusProps)}
          ref={ref ? mergeRefs(ref, localRef) : localRef}
          data-id={dataId ? `${dataId}--input` : undefined}
        />
      </span>
    );

    if (groupState) {
      if (props.value == null) {
        console.warn(
          "A <Checkbox> element within a <CheckboxGroup> requires a `value` property."
        );
      }
    }

    return (
      <>
        {children ? (
          <Label
            className={classNames}
            {...{ inputControl, inputPosition }}
            visuallyHidden={visuallyHideLabel}
            data-id={dataId ? `${dataId}--label` : undefined}
          >
            {children}
          </Label>
        ) : (
          <span
            className={classNames}
            data-id={dataId ? `${dataId}--noLabel` : undefined}
          >
            {inputControl}
          </span>
        )}
      </>
    );
  }
);

export default Checkbox;

Checkbox.displayName = "Checkbox";
