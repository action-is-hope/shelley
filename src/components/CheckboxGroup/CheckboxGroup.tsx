import { Ref, forwardRef, ReactElement } from "react";
import type { FieldProps } from "../Field/Field";
import type { AriaCheckboxGroupProps } from "@react-types/checkbox";
import type { CheckboxProps } from "../Checkbox/Checkbox";
import type { Orientation } from "src/typings/shared-types";
import { useCheckboxGroupState } from "react-stately";
import { useCheckboxGroup } from "react-aria";
import { CheckboxGroupContext } from "./context";
import { Field } from "../Field/Field";
/* = Style API. */
import { st, classes } from "./checkboxGroup.st.css";

export interface CheckboxGroupProps
  extends Omit<AriaCheckboxGroupProps, "excludeFromTabOrder">,
    Omit<
      FieldProps,
      | "label"
      | "startAdornment"
      | "endAdornment"
      | "disableLabelTransition"
      | "hasValue"
      | "variant"
    > {
  /**
   * Position of the label.
   * @default "over"
   */
  labelPosition?: "top" | "side" | "hidden";
  className?: string;
  /**
   * The Checkboxes contained within the CheckboxGroup.
   */
  children: ReactElement<CheckboxProps> | ReactElement<CheckboxProps>[];
  /**
   * The axis the checkboxes should align with.
   * @default 'vertical'
   */
  orientation?: Orientation;
}

function CheckboxGroup(props: CheckboxGroupProps, ref?: Ref<HTMLDivElement>) {
  const {
    className: classNameProp,
    description,
    isDisabled,
    errorMessage,
    validationState,
    label,
    labelPosition = "top",
    orientation = "vertical",
    vol,
    children,
    "data-id": dataId,
  } = props;

  const state = useCheckboxGroupState(props);
  const { labelProps, groupProps, descriptionProps, errorMessageProps } =
    useCheckboxGroup(props, state);

  return (
    <Field
      {...{
        isDisabled,
        errorMessage,
        validationState,
        errorMessageProps,
        description,
        descriptionProps,
        label,
        labelPosition,
        labelProps,
        vol,
        variant: false,
        fieldContainerProps: { ...groupProps, className: classes.group },
        ref,
        "data-id": dataId,
      }}
      className={st(
        classes.root,
        {
          orientation,
        },
        classNameProp
      )}
    >
      <>
        <CheckboxGroupContext.Provider value={state}>
          {children}
        </CheckboxGroupContext.Provider>
      </>
    </Field>
  );
}

const _CheckboxGroup = forwardRef(CheckboxGroup);
export { _CheckboxGroup as CheckboxGroup };
