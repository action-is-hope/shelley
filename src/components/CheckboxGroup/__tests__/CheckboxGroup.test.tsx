import { CheckboxGroup } from "../CheckboxGroup";
import Checkbox from "../../Checkbox/Checkbox";
import renderer from "react-test-renderer";

describe("Checkbox", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <CheckboxGroup label="Field Label">
          <Checkbox value="cb1">Label1</Checkbox>
          <Checkbox value="cb2">Label2</Checkbox>
        </CheckboxGroup>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with custom labelled checkboxes", () => {
    const tree = renderer
      .create(
        <CheckboxGroup label="Field Label">
          <label htmlFor="cb1">Label1</label>
          <Checkbox id="cb1" />
          <label htmlFor="cb2">Label2</label>
          <Checkbox id="cb2" />
        </CheckboxGroup>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
