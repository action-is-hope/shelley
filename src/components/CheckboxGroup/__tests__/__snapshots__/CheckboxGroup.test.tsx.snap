// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
<div
  className="Field__Test__root Field__Test---vol-1-1 Field__Test---labelPosition-3-top CheckboxGroup__Test__root CheckboxGroup__Test---orientation-8-vertical"
>
  <label
    className="Label__Test__root Field__Test__inputLabel"
    id="react-aria-1"
  >
    <span
      className="Label__Test__labelText"
    >
      Field Label
    </span>
  </label>
  <div
    aria-labelledby="react-aria-1"
    className="Field__Test__fieldContainer CheckboxGroup__Test__group"
    id="cbg1"
    role="group"
  >
    <label
      className="Label__Test__root Label__Test--hasInput Checkbox__Test__root Checkbox__Test---vol-1-1"
    >
      <span
        className="Label__Test__labelText"
      >
        Label1
      </span>
      <span
        className="Checkbox__Test__inputContainer"
      >
        <input
          checked={false}
          className="Checkbox__Test__input"
          disabled={false}
          id="cb1"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          type="checkbox"
          value="cb1"
        />
      </span>
    </label>
    <label
      className="Label__Test__root Label__Test--hasInput Checkbox__Test__root Checkbox__Test---vol-1-1"
    >
      <span
        className="Label__Test__labelText"
      >
        Label2
      </span>
      <span
        className="Checkbox__Test__inputContainer"
      >
        <input
          checked={false}
          className="Checkbox__Test__input"
          disabled={false}
          id="cb2"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          type="checkbox"
          value="cb2"
        />
      </span>
    </label>
  </div>
</div>
`;

exports[`Checkbox renders correctly with custom labelled checkboxes 1`] = `
<div
  className="Field__Test__root Field__Test---vol-1-1 Field__Test---labelPosition-3-top CheckboxGroup__Test__root CheckboxGroup__Test---orientation-8-vertical"
>
  <label
    className="Label__Test__root Field__Test__inputLabel"
    id="react-aria-1"
  >
    <span
      className="Label__Test__labelText"
    >
      Field Label
    </span>
  </label>
  <div
    aria-labelledby="react-aria-1"
    className="Field__Test__fieldContainer CheckboxGroup__Test__group"
    id="cbg1"
    role="group"
  >
    <label
      htmlFor="cb1"
    >
      Label1
    </label>
    <span
      className="Checkbox__Test__root Checkbox__Test---vol-1-1"
    >
      <span
        className="Checkbox__Test__inputContainer"
      >
        <input
          checked={false}
          className="Checkbox__Test__input"
          disabled={false}
          id="cb1"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          type="checkbox"
          value="cb1"
        />
      </span>
    </span>
    <label
      htmlFor="cb2"
    >
      Label2
    </label>
    <span
      className="Checkbox__Test__root Checkbox__Test---vol-1-1"
    >
      <span
        className="Checkbox__Test__inputContainer"
      >
        <input
          checked={false}
          className="Checkbox__Test__input"
          disabled={false}
          id="cb2"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          type="checkbox"
          value="cb2"
        />
      </span>
    </span>
  </div>
</div>
`;
